src/Voron/Data/Tables/AbstractSchemaIndexDefinition.cs

```diff
@@ -0,0 +1,43 @@
+using Sparrow;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using Sparrow.Binary;
+using Sparrow.Server;
+using Voron.Data.BTrees;
+using Voron.Data.RawData;
+using Voron.Impl;
+
+namespace Voron.Data.Tables
+{
+    [Flags]
+    public enum TableIndexType
+    {
+        Default = 0x01,
+        BTree = 0x01,
+        Custom = 0x2
+    }
+
+    public abstract class AbstractSchemaIndexDefinition
+    {
+        public abstract TableIndexType Type { get;}
+
+        public bool IsGlobal;
+
+        public Slice Name;
+
+        public abstract ByteStringContext.Scope GetSlice(ByteStringContext context, ref TableValueReader value,
```

Comment:
These are heavy duty calls because of the number of calls. Going abstract here without ensuring devirtualization happens at the JIT level is very risky. We should ensure that it does not regress insertation/mutation/deletion performance.

Reply:
We can apply some of the techniques we created for Corax to solve this type of conundrum, but I will be working on that further down the line with a rework of the performance edge of Voron itself that we have already scheduled with @ayende to happen during the 6.0 development timeframe or even 6.1; I would hesitate a bit to introduce those techniques to solve this before we do all the required work done to push into that direction.

Reply:
We have a plan to do some perf tests - https://issues.hibernatingrhinos.com/issue/RavenDB-18076.

But if you think it'd be better / safer to get rid of abstract here then we might adjust that I think. @aviv86 ?

Reply:
I would say, do them first... lets ensure that we are not regressing hard before continuing.

Reply:
Yes, that's the main point of doing it.

Reply:
> These are heavy duty calls because of the number of calls. Going abstract here without ensuring devirtualization happens at the JIT level is very risky. We should ensure that it does not regress insertion/mutation/deletion performance. During replication/sharding it may not be that different as those are usually network dominated operations, you have to time this against local operations.

what if instead of inheriting from an abstract class, index definitions will implement a common interface - will it make any difference in this aspect? less risky?

Reply:
do we have any existing benchmarks for those?

Reply:
> do we have any existing benchmarks for those?
yes we have a few benchmarks here : 
https://github.com/ravendb/ravendb/tree/v5.3/bench/Voron.Benchmark/Table

Reply:
>> what if instead of inheriting from an abstract class, index definitions will implement a common interface - will it make any difference in this aspect? less risky?
Not really the problem is that the calls become virtual, and at heavy duty places that essentially kills you.

Reply:
@redknightlois I intend to sit with @aviv86 next week and change that to generic struct interface to avoid the virtual dispatch, yes. This is done along the way to drive the impl.

Reply:
@aviv86 Can you post the before & after results of those benchmarks?

Reply:
> @aviv86 Can you post the before & after results of those benchmarks?

https://issues.hibernatingrhinos.com/issue/RavenDB-18076#focus=Comments-67-274182.0-0

Reply:
made some [changes](https://github.com/ravendb/ravendb/pull/13648/commits/0d6ae3790d1d2896e45e28a339f969d021252f68)
- removed `AbstractTableIndexDef` 
- removed all abstract and virtual calls from `AbstractTreeIndexDef` 
- `AbstractTreeIndexDef` now contains only the common properties (Name, IsGlobal, Type) and is used only in "Seek" methods, e.g. : `SeekForwardFromPrefix(AbstractTreeIndexDef index, ...)`,  `SeekBackwardFrom(AbstractTreeIndexDef  index, ...)`, etc.